File: shared/taskStatus.ts
import type {
  SubtaskProgress,
  SubtaskProgressInput,
  Task,
  TaskEffort,
  TaskI18nKeys,
  TaskPriority,
  TaskStatus,
  TaskUser,
} from './types';

/**
 * Per-status glyph for the on-chip / on-row marker.
 *
 * Glyph-only — no SVG, no font dependency, no row-shift surprises across
 * locales. All four glyphs sit on a common visual spine (the circle) so the
 * user reads them as a progression rather than four unrelated symbols.
 *
 *   open        ○  empty circle (U+25CB)
 *   in_progress ◐  half-filled circle (U+25D0) — "started, not finished"
 *   completed   ●  filled circle (U+25CF)
 *   cancelled   ⊘  slashed circle — "abandoned / no longer pursued"
 *
 * All four are drawn from the same Geometric-Shapes family at a matching
 * nominal size so they read as ONE progression. `completed` uses U+25CF
 * (BLACK CIRCLE), the size-matched fill of the ○/◐ outline — NOT U+2B24
 * (BLACK LARGE CIRCLE), which renders visibly oversized against the others
 * on the mobile system fonts.
 *
 * Shared across every task surface so the symbols mean the same thing wherever
 * the user sees them.
 */
export function statusMarker(status: TaskStatus): string {
  switch (status) {
    case 'completed':
      return '●';
    case 'cancelled':
      return '⊘';
    case 'in_progress':
      return '◐';
    case 'open':
      return '○';
  }
}

/**
 * How many levels the user's priority system has — the synced
 * `tasks.twoLevelPriority` setting (Settings → Tasks), as a value rather than a
 * bare boolean so every reader says what it means.
 *
 *   - `'three'` — low / medium / high, the original system.
 *   - `'two'` — normal / important. Everything that is not the top level is
 *     simply normal, carries no indicator at all, and sorts in one band with
 *     the rest of normal. Modelled on a bullet journal, where the only mark a
 *     task gets is the star that says "this one".
 *
 * The STORED value is untouched by the setting. A task that arrived from a
 * provider as `low` stays `low` in the database and on the provider — the
 * two-level system is a lens over the same three values, so switching back and
 * forth loses nothing and no provider's data is rewritten behind the user's
 * back. That is also what makes the "collapse everything below the top" rule
 * safe for providers with five or six levels: their extra levels were already
 * folded into these three by their adapter, and this folds two of the three
 * into one for reading.
 */
export type { PriorityScale } from './generated/PriorityScale';
import type { PriorityScale } from './generated/PriorityScale';

/**
 * The priority RULES — installed by the surface, not implemented here.
 *
 * They live in `cal_core::task_priority`. What differs per surface is only how
 * Rust is reached: the desktop through WebAssembly compiled into its webview,
 * mobile through a synchronous Expo `Function`, a future native frontend by
 * linking it. Each installs its own door at startup and everything in this
 * package then ranks the same way.
 *
 * They were written here first, then copied into `crates/cal-core-wasm` when
 * the desktop needed a synchronous road into Rust — which left the rule
 * reachable from ONE surface while the other two ran this file. Two devices
 * showing one task list have to put the same task first, so the copy is gone
 * and this is the door.
 *
 * Note what did NOT move: the glyphs. The core answers with an order, a state
 * or a KEY; each surface picks its own characters, because an e-ink display
 * plausibly wants different ones.
 */
export interface TaskPriorityRules {
  /** Sort rank; 0 sorts first. */
  priorityRank(priority: TaskPriority, scale: PriorityScale): number;
  /** The TOP priority — "important" in the two-level system. */
  isImportantPriority(priority: TaskPriority): boolean;
  /** What "important" clears to: what the task had, unless that was the top. */
  normalPriority(previous?: TaskPriority | null): TaskPriority;
}

let installedPriority: TaskPriorityRules | null = null;

/** Bind this surface's door into the core. */
export function installTaskPriorityRules(rules: TaskPriorityRules): void {
  installedPriority = rules;
}

function priorityRules(): TaskPriorityRules {
  if (installedPriority === null) {
    // Loud, not a local fallback. A fallback here would be a fourth copy of
    // the ranking, and the failure it produces — one device ordering a list
    // differently from another — is exactly what nobody reports and everybody
    // trips over.
    throw new Error(
      'task priority rules used before installTaskPriorityRules() — the ' +
        'surface must install its door into cal-core during startup',
    );
  }
  return installedPriority;
}

/**
 * Whether the task carries the TOP priority — the one level that survives in
 * the two-level system, where it is called "important" rather than "high".
 */
export function isImportantPriority(priority: TaskPriority): boolean {
  return priorityRules().isImportantPriority(priority);
}

/**
 * The priority a task gets when the user clears "important" in the two-level
 * system: the value it already had, as long as that value is not the top one.
 *
 * Unchecking must not rewrite `low` into `medium`. Both read as "normal" and
 * both look identical on every surface, so the write would change nothing the
 * user can see while changing what other clients — and the three-level system,
 * if they switch back — display. `previous` is what the task carried before it
 * was marked important; `medium` is the neutral answer when there is none.
 */
export function normalPriority(previous?: TaskPriority | null): TaskPriority {
  return priorityRules().normalPriority(previous);
}

/**
 * Per-priority glyph for the on-chip / on-row indicator.
 *
 * Three-level: exclamation marks, one per level — `!` low, `!!` medium, `!!!`
 * high. Two-level: a star on the important one and NOTHING on the rest, which
 * is the point of the second system — normal is the absence of a mark, not a
 * quieter mark.
 *
 * The glyph carries the meaning; the SR label still spells the priority out via
 * {@link prioritySuffix}. `scale` is required rather than defaulted so that
 * adding a surface cannot silently print exclamation marks to a user who turned
 * them off.
 */
export function priorityMarker(
  priority: TaskPriority,
  scale: PriorityScale,
): string {
  if (scale === 'two') return isImportantPriority(priority) ? '★' : '';
  switch (priority) {
    case 'high':
      return '!!!';
    case 'low':
      return '!';
    case 'medium':
      return '!!';
  }
}

/**
 * Numeric sort rank for a priority (ascending = most urgent first): `high` → 0,
 * `medium` → 1, `low` → 2. Pair with a *stable* sort so the existing order is
 * the tiebreaker within one priority bucket.
 *
 * Two-level collapses low and medium into ONE band (important → 0, normal → 1).
 * It has to: the two are indistinguishable on screen there, so ranking them
 * apart would order a list by an attribute the reader cannot perceive, and the
 * A→Z tiebreak that each band promises would appear to break at random.
 *
 * `scale` defaults to `'three'`, the historical behaviour, because this feeds
 * comparators handed straight to `Array.sort` — a caller that omits it sorts
 * exactly as before rather than failing to compile.
 */
export function priorityRank(
  priority: TaskPriority,
  // Defaulted here rather than in the core, because the default is about THIS
  // boundary: the callers are comparators handed straight to `Array.sort`, and
  // one that omits the scale should sort exactly as it always did rather than
  // fail to compile.
  scale: PriorityScale = 'three',
): number {
  return priorityRules().priorityRank(priority, scale);
}

// ─────────────────── The state door: keys and progress ──────────────────────

/**
 * This surface's door into `cal_core::task_status`.
 *
 * Two answers cross it. The i18n KEY every state, effort and priority is
 * announced with — as ONE table, because the views ask per chip and on the
 * phone every ask is a trip across the native bridge for a word that never
 * changes while the app runs; the core publishes the whole table and this
 * shell reads it once. And how far a parent's subtasks are, answered for every
 * parent of a list in one crossing, remembered here per task array because
 * the views ask per row.
 *
 * Pinned by `crates/cal-core/tests/fixtures/taskStatus.json`, measured from
 * the TypeScript this replaced; `taskStatus.contract.test.ts` replays it
 * through this door, and the core's own contract test reads the same file.
 */
export interface TaskStatusRules {
  /** The whole key table, `cal_core::task_status::TaskI18nKeys`. */
  taskI18nKeysJson(): string;
  /** Parent id → `{done, total}` for every parent with children that count. */
  subtaskProgressJson(inputJson: string): string;
}

let installedStatus: TaskStatusRules | null = null;
let keyTable: TaskI18nKeys | null = null;
let progressByList: WeakMap<readonly Task[], Record<string, SubtaskProgress>> =
  new WeakMap();

/** Bind this surface's door into the core. */
export function installTaskStatusRules(rules: TaskStatusRules): void {
  installedStatus = rules;
  // A re-install (tests) must not serve the previous door's answers.
  keyTable = null;
  progressByList = new WeakMap();
}

function statusRules(): TaskStatusRules {
  if (installedStatus === null) {
    // Loud, not a local fallback. A fallback here would be the TypeScript
    // twin all over again, and its failure is one a screen-reader user meets
    // head on: one device announcing a state the other one does not.
    throw new Error(
      'task status rules used before installTaskStatusRules() — the surface ' +
        'must bind its door into cal_core::task_status at startup',
    );
  }
  return installedStatus;
}

/** The key table, read through the door on the first ask and kept. */
function keys(): TaskI18nKeys {
  if (keyTable === null) {
    keyTable = JSON.parse(statusRules().taskI18nKeysJson()) as TaskI18nKeys;
  }
  return keyTable;
}

/**
 * i18n key for the SR-announced priority label, or `null` when there is nothing
 * to announce — `medium` in the three-level system, and everything below the
 * top in the two-level one.
 *
 * Two-level says "important", not "high priority": in a system with one mark,
 * naming a level the user cannot choose between would describe a scale that is
 * no longer there.
 */
export function priorityI18nKey(
  priority: TaskPriority,
  scale: PriorityScale,
): string | null {
  return keys().priority[scale][priority];
}

/**
 * SR-friendly priority suffix for aria-labels. Empty string when there is
 * nothing to say, so it can be appended unconditionally to any task label.
 * Comma-prefixed, so the calling i18n template
 * (`{{state}}{{priority}}{{progress}}`) needs no conditional.
 */
export function prioritySuffix(
  t: (key: string, vars?: Record<string, unknown>) => string,
  priority: TaskPriority,
  scale: PriorityScale,
): string {
  const key = priorityI18nKey(priority, scale);
  return key ? `, ${t(key)}` : '';
}

/**
 * i18n key for the SR-announced effort label, or `null` for `medium` (the
 * neutral default — no announcement, mirroring `priorityI18nKey`).
 */
export function effortI18nKey(effort: TaskEffort): string | null {
  return keys().effort[effort];
}

/**
 * SR-friendly effort suffix for aria-labels (", großer Aufwand"). Empty string
 * for `medium` so it appends unconditionally. Comma-prefixed like
 * {@link prioritySuffix}. Always present regardless of the visual-sizing toggle,
 * so a screen-reader user always hears the effort.
 */
export function effortSuffix(
  t: (key: string, vars?: Record<string, unknown>) => string,
  effort: TaskEffort,
): string {
  const key = effortI18nKey(effort);
  return key ? `, ${t(key)}` : '';
}

/**
 * The chip-class modifier token for an effort's visual tile size, or '' for
 * `small` — the whole scale sits one step above the original mapping (tester
 * feedback): small renders at the neutral base size (no class), medium at the
 * former "large" size, large bigger still. Single source of truth for the
 * effort→size hook: a caller builds `${prefix}--effort-${token}` only when
 * the token is non-empty AND the user's `visualEffortSizing` pref is on.
 */
export function effortSizeModifier(effort: TaskEffort): string {
  return effort === 'small' ? '' : effort;
}

/**
 * i18n key for the SR-announced state suffix. Keys live under
 * `views.tasks.state*`. Adding a new TaskStatus value? The core's match is
 * exhaustive, the published table grows a field, and this lookup stops
 * compiling until the generated `StatusKeys` carries it — plus the locale
 * files.
 */
export function statusI18nKey(status: TaskStatus): string {
  return keys().status[status];
}

/**
 * Count completed children of `parentId` among `allTasks`. Returns `null` when
 * the parent has no children at all. "Done" means `completed`; `cancelled`
 * rows drop out of the total so the fraction reflects what's left to do.
 *
 * One crossing per task ARRAY, not per row: the core answers for every parent
 * of the list at once, and the answer is kept as long as the array lives —
 * the views hand the same list to every row they render.
 */
export function subtaskProgress(
  parentId: string,
  allTasks: readonly Task[],
): { done: number; total: number } | null {
  let byParent = progressByList.get(allTasks);
  if (byParent === undefined) {
    // Only what the count reads crosses; a task carries far more.
    const input: SubtaskProgressInput = {
      tasks: allTasks.map((t) => ({
        id: t.id,
        status: t.status,
        parent_id: t.parent_id ?? null,
      })),
    };
    byParent = JSON.parse(
      statusRules().subtaskProgressJson(JSON.stringify(input)),
    ) as Record<string, SubtaskProgress>;
    progressByList.set(allTasks, byParent);
  }
  const progress = byParent[parentId];
  return progress === undefined ? null : progress;
}

/**
 * SR-friendly progress suffix for aria-labels. Empty string when the task has
 * no children. Resolves through `views.tasks.subtaskProgress` which already
 * starts with a comma separator.
 */
export function subtaskProgressSuffix(
  t: (key: string, vars?: Record<string, unknown>) => string,
  parentId: string,
  allTasks: readonly Task[],
): string {
  const progress = subtaskProgress(parentId, allTasks);
  if (!progress) return '';
  return t('views.tasks.subtaskProgress', progress);
}

/**
 * The parent task's title when `task` is a subtask whose parent is present in
 * `allTasks`, else null. Used to label a subtask chip in context wherever it
 * surfaces on its own (calendar grid, backlog rail).
 */
export function subtaskParentTitle(
  task: Task,
  allTasks: Task[],
): string | null {
  if (!task.parent_id) return null;
  return allTasks.find((p) => p.id === task.parent_id)?.title ?? null;
}

/**
 * SR-friendly suffix naming a subtask's parent (", Unteraufgabe von X"), or ''
 * when the task isn't a subtask. Resolves through `views.tasks.subtaskParent`,
 * which starts with a comma separator — mirrors {@link subtaskProgressSuffix}.
 */
export function subtaskParentSuffix(
  t: (key: string, vars?: Record<string, unknown>) => string,
  task: Task,
  allTasks: Task[],
): string {
  const parent = subtaskParentTitle(task, allTasks);
  return parent ? t('views.tasks.subtaskParent', { parent }) : '';
}

/**
 * SR-friendly assignee suffix for aria-labels. Empty string when the task has
 * no assignees. Resolves through `views.tasks.assigneeSuffix`, which starts
 * with a comma separator. Shared across every task surface so an assignment is
 * announced wherever the task appears.
 */
export function assigneeSuffix(
  t: (key: string, vars?: Record<string, unknown>) => string,
  assignees: TaskUser[],
): string {
  if (assignees.length === 0) return '';
  return t('views.tasks.assigneeSuffix', {
    names: assignees.map((a) => a.name).join(', '),
  });
}

File: crates/cal-core/src/task_status.rs
//! What a task's state, effort and priority are announced with, and how far a
//! parent's subtasks are.
//!
//! Two of the answers `shared/taskStatus.ts` used to give in TypeScript. What
//! did NOT move is deliberate, and it is the decision of 2026-09-09: the core
//! answers a STATE or a KEY, never a character and never a sentence. The
//! glyphs (○ ◐ ● ⊘, ! !! !!! and ★) stay on each surface — an e-ink display
//! plausibly wants others — and so does every suffix built with `t`. What
//! moved is what every surface has to agree on: which i18n key a value is
//! announced with (DESIGN §4.5 a; `ConferenceProvider::i18n_key` is the
//! precedent), and the one arithmetic in the module.
//!
//! # Published once, not asked per chip
//!
//! The keys are constants, and the views ask for them per chip. A door per
//! call would cross the phone's native bridge once per row to learn a word it
//! could have read at startup, so the core publishes the whole table as
//! [`TaskI18nKeys`] and a surface reads it once when it installs its door.
//!
//! # Progress, for every parent at once
//!
//! The views ask "how far is this parent" per row, over the same task list.
//! The core answers for every parent in one crossing; the shell remembers the
//! answer per task array.
//!
//! Every answer here is pinned by `tests/fixtures/taskStatus.json`, measured by
//! running the TypeScript this replaces; the `contract` module below reads it,
//! and so does the TypeScript contract test on the other side of the boundary.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::task_priority::PriorityScale;
use crate::types::{TaskEffort, TaskPriority, TaskStatus};

/// The i18n key a task's state is announced with. Every state has one.
pub fn status_i18n_key(status: TaskStatus) -> &'static str {
    match status {
        TaskStatus::Open => "views.tasks.stateOpen",
        TaskStatus::InProgress => "views.tasks.stateInProgress",
        TaskStatus::Completed => "views.tasks.stateDone",
        TaskStatus::Cancelled => "views.tasks.stateCancelled",
    }
}

/// The i18n key a task's effort is announced with, or `None` for the neutral
/// middle — nothing to announce.
pub fn effort_i18n_key(effort: TaskEffort) -> Option<&'static str> {
    match effort {
        TaskEffort::Small => Some("views.tasks.effortSmall"),
        TaskEffort::Medium => None,
        TaskEffort::Large => Some("views.tasks.effortLarge"),
    }
}

/// The i18n key a task's priority is announced with under `scale`, or `None`
/// when there is nothing to announce: the middle of the three-level system,
/// and everything below the top in the two-level one. Two-level says
/// "important", not "high priority": in a system with one mark, naming a level
/// the user cannot choose between would describe a scale that is no longer
/// there.
pub fn priority_i18n_key(priority: TaskPriority, scale: PriorityScale) -> Option<&'static str> {
    match scale {
        PriorityScale::Two => priority
            .is_important()
            .then_some("views.tasks.priorityImportant"),
        PriorityScale::Three => match priority {
            TaskPriority::High => Some("views.tasks.priorityHigh"),
            TaskPriority::Medium => None,
            TaskPriority::Low => Some("views.tasks.priorityLow"),
        },
    }
}

/// The keys for every state.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct StatusKeys {
    pub open: String,
    pub in_progress: String,
    pub completed: String,
    pub cancelled: String,
}

/// The keys for every effort; `None` where nothing is announced.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct EffortKeys {
    pub small: Option<String>,
    pub medium: Option<String>,
    pub large: Option<String>,
}

/// The keys for every priority under one scale; `None` where nothing is
/// announced.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct PriorityKeys {
    pub low: Option<String>,
    pub medium: Option<String>,
    pub high: Option<String>,
}

/// The priority keys per scale.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct PriorityKeysByScale {
    pub three: PriorityKeys,
    pub two: PriorityKeys,
}

/// Every i18n key of the task vocabulary, as one table — what a surface reads
/// once when it installs its door.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct TaskI18nKeys {
    pub status: StatusKeys,
    pub effort: EffortKeys,
    pub priority: PriorityKeysByScale,
}

fn priority_keys(scale: PriorityScale) -> PriorityKeys {
    let key = |p| priority_i18n_key(p, scale).map(str::to_string);
    PriorityKeys {
        low: key(TaskPriority::Low),
        medium: key(TaskPriority::Medium),
        high: key(TaskPriority::High),
    }
}

/// The whole table. See the module doc for why it is a table.
pub fn task_i18n_keys() -> TaskI18nKeys {
    TaskI18nKeys {
        status: StatusKeys {
            open: status_i18n_key(TaskStatus::Open).to_string(),
            in_progress: status_i18n_key(TaskStatus::InProgress).to_string(),
            completed: status_i18n_key(TaskStatus::Completed).to_string(),
            cancelled: status_i18n_key(TaskStatus::Cancelled).to_string(),
        },
        effort: EffortKeys {
            small: effort_i18n_key(TaskEffort::Small).map(str::to_string),
            medium: effort_i18n_key(TaskEffort::Medium).map(str::to_string),
            large: effort_i18n_key(TaskEffort::Large).map(str::to_string),
        },
        priority: PriorityKeysByScale {
            three: priority_keys(PriorityScale::Three),
            two: priority_keys(PriorityScale::Two),
        },
    }
}

/// [`task_i18n_keys`] over the wire.
pub fn task_i18n_keys_json() -> String {
    serde_json::to_string(&task_i18n_keys()).expect("a table of strings serializes")
}

/// What the progress reads of a task — by the same field names as `Task`, so
/// a full task's JSON deserializes into it.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct ProgressTask {
    pub id: String,
    pub status: TaskStatus,
    #[serde(default)]
    pub parent_id: Option<String>,
}

/// The question [`subtask_progress`] answers, over the wire.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct SubtaskProgressInput {
    pub tasks: Vec<ProgressTask>,
}

/// How far a parent's subtasks are: completed direct children over direct
/// children that are not cancelled.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-export", derive(ts_rs::TS), ts(export))]
pub struct SubtaskProgress {
    pub done: usize,
    pub total: usize,
}

/// Parent id → progress, for every parent that has children that count. A
/// parent absent from the answer has no children, or only cancelled ones:
/// nothing left to do and nothing done, no fraction to speak. Only DIRECT
/// children count — a grandchild is its own parent's business. Whether the
/// parent's own row is in the list does not matter; the count is by
/// `parent_id`.
pub fn subtask_progress(tasks: &[ProgressTask]) -> BTreeMap<String, SubtaskProgress> {
    let mut out: BTreeMap<String, SubtaskProgress> = BTreeMap::new();
    for task in tasks {
        let Some(parent) = task.parent_id.as_deref() else {
            continue;
        };
        if task.status == TaskStatus::Cancelled {
            continue;
        }
        let progress = out
            .entry(parent.to_string())
            .or_insert(SubtaskProgress { done: 0, total: 0 });
        progress.total += 1;
        if task.status == TaskStatus::Completed {
            progress.done += 1;
        }
    }
    out
}

/// [`subtask_progress`] over the wire.
pub fn subtask_progress_json(input_json: &str) -> Result<String, serde_json::Error> {
    let input: SubtaskProgressInput = serde_json::from_str(input_json)?;
    serde_json::to_string(&subtask_progress(&input.tasks))
}

/// The contract with the TypeScript this replaced.
///
/// Its other half is `src/intl/taskStatus.contract.test.ts`, reading this
/// same file. The fixture was written by running the TypeScript BEFORE the
/// port, so the port is measured against what was.
#[cfg(test)]
mod contract {
    use super::*;
    use serde_json::{json, Value};

    const CONTRACT: &str = include_str!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/tests/fixtures/taskStatus.json"
    ));

    #[test]
    fn every_key_in_the_table_holds() {
        let doc: Value = serde_json::from_str(CONTRACT).expect("the contract parses");
        // Anti-silence: the rows the rules turn on, named.
        assert_eq!(
            doc["keys"]["priority"]["two"]["high"],
            "views.tasks.priorityImportant"
        );
        assert_eq!(doc["keys"]["priority"]["two"]["low"], Value::Null);
        assert_eq!(doc["keys"]["priority"]["three"]["medium"], Value::Null);
        assert_eq!(doc["keys"]["effort"]["medium"], Value::Null);
        let got = serde_json::to_value(task_i18n_keys()).expect("the table serializes");
        assert_eq!(got, doc["keys"]);
    }

    #[test]
    fn every_progress_case_holds() {
        let doc: Value = serde_json::from_str(CONTRACT).expect("the contract parses");
        let cases = doc["progress"].as_array().expect("progress is an array");
        for needed in [
            "counts-completed-and-drops-cancelled-from-the-total",
            "direct-children-only",
            "only-cancelled-children-is-null",
        ] {
            assert!(
                cases.iter().any(|c| c["name"] == needed),
                "the contract lost `{needed}`"
            );
        }
        let base = &doc["baseTask"];
        for case in cases {
            let name = case["name"].as_str().expect("every case has a name");
            let tasks: Vec<ProgressTask> = case["input"]["tasks"]
                .as_array()
                .expect("tasks is an array")
                .iter()
                .map(|over| {
                    let mut row = base.clone();
                    for (k, v) in over.as_object().expect("an override object") {
                        row[k] = v.clone();
                    }
                    serde_json::from_value(row).expect("a task row deserializes")
                })
                .collect();
            let answer = subtask_progress(&tasks);
            let mut got = serde_json::Map::new();
            for parent in case["input"]["parents"]
                .as_array()
                .expect("parents is an array")
            {
                let parent = parent.as_str().expect("a parent id");
                got.insert(
                    parent.to_string(),
                    answer
                        .get(parent)
                        .map(|p| json!({ "done": p.done, "total": p.total }))
                        .unwrap_or(Value::Null),
                );
            }
            assert_eq!(
                Value::Object(got),
                case["expect"],
                "{name}: {}",
                case["note"].as_str().unwrap_or("")
            );
        }
    }

    #[test]
    fn the_wire_accepts_a_full_task() {
        let input = r#"{"tasks": [
            {"id": "p", "list_id": "L", "title": "P", "status": "open", "priority": "medium",
             "effort": "medium", "parent_id": null, "assignees": [], "reminders": [],
             "created_at": "2026-01-01T00:00:00Z", "updated_at": "2026-01-01T00:00:00Z"},
            {"id": "c", "list_id": "L", "title": "C", "status": "completed", "priority": "medium",
             "parent_id": "p"}
        ]}"#;
        let answer: BTreeMap<String, SubtaskProgress> =
            serde_json::from_str(&subtask_progress_json(input).expect("the input is valid"))
                .expect("the answer round-trips");
        assert_eq!(answer["p"], SubtaskProgress { done: 1, total: 1 });
        assert!(!task_i18n_keys_json().is_empty());
    }
}
